aeppl/joint_logprob.py

```diff
@@ -213,6 +213,20 @@ def factorized_joint_logprob(
             for node in io_toposort(graph_inputs(q_logprob_vars), q_logprob_vars):
                 compute_test_value(node)
 
+    missing_value_terms = set(original_values.values()) - set(logprob_vars.keys())
+    if missing_value_terms:
+        # Filter out missing terms of variables with ignore_logprob
+        value_rvs = {v: k for k, v in updated_rv_values.items()}
+        for missing in tuple(missing_value_terms):
+            rv_of_missing = value_rvs.get(missing, None)
+            if rv_of_missing and getattr(rv_of_missing.tag, "ignore_logprob", False):
+                missing_value_terms.remove(missing)
```

Comment:
> When I said something could be removed later, I meant the specific ignore_logprob flag logic, not the raising a RuntimeError if a variable is missing.
> 
> While developing the censored variables it would often fail silently and just return a graph with aesara clips unchanged and/or less terms than requested. This seems like a good way to catch such failures.

From a simple development and design perspective, it sounds like you're addressing a testing-specific issue within a feature implementation, and that's generally not good.

Otherwise, if something is failing silently the first question is "What's failing?".  Is it the `factorized_joint_logprob` loop?  If not, the failure should be addressed closer to where its primary logic/code resides, and that doesn't appear to be here.

> Is there a reason why you don't want this type of check at the end of `factorized_joint_logprob`?
> 
> It's trivial to manually do the same check in those new tests I added, but this explicit check might be valuable enough to have as a default.

The reason why I don't want these kinds of unrelated changes is that their inclusion makes a PR contingent on additional review work and discussions.

It takes extra time and effort to go through logic like this and determine its relevance, risk, etc.  These are things that need to be done within issues and/or at the outset of a PR (e.g. the premise/description of a PR) in order to avoid delaying the inclusion of any agreed upon changes.

Reply:
> From a simple development and design perspective, it sounds like you're addressing a testing-specific issue within a feature implementation, and that's generally not good.

What happened was that the test revealed that such a call to factorized_logprob  would return with a missing logp term and zero complaints so I decided to add an explicit check there. 

It was not for the sake of the test as that had already been solved and could be tested explicitly inside the test itself.

It was meant for further development when we introduce rewrites for Ops that are otherwise valid in logp graphs. It's also a conceptual obvious check for me: a user requested a dictionary of rv_values and we make sure we are returning a dictionary with a item for each original pair.

I don't mind splitting this into another PR

Reply:
I removed the commit but I am not very happy with the fact that this does not fail naturally:

```python
def test_fail_base_and_censored_have_values():
    """Test failure when both base_rv and clipped_rv are given value vars"""
    x_rv = at.random.normal(0, 1)
    cens_x_rv = at.clip(x_rv, x_rv, 1)
    cens_x_rv.name = "cens_x"

    x_vv = x_rv.clone()
    cens_x_vv = cens_x_rv.clone()
    logp_terms = factorized_joint_logprob({cens_x_rv: cens_x_vv, x_rv: x_vv})
    assert cens_x_vv not in logp_terms
```

Reply:
Again, why should the error be raised in `factorized_joint_logprob` specifically?  

The two terms involved are very specific to the censored variable logic, and it looks like the error could've been initiated in `find_censored_rvs`&mdash;i.e. where all the relevant terms are identified and used.  This approach could also short-circuit all the unnecessary down-stream logic, no?

Reply:
I don't think it should be the rewrite responsibility to raise a failure. It may be there is another rewrite (e.g., added by users of the library) that can handle the conversion.

The bigger problem is that nothing happens if you ask for a graph that we don't know how to handle. That is not specific to censoredRVs, we just haven't tested it. For instance, this snippet does not complain at all:

```python
import aesara.tensor as at
import aeppl

x_rv = at.random.normal(name='x')
y_rv = at.cos(x_rv)

x_vv = x_rv.clone()
y_vv = y_rv.clone()

logprob_dict = aeppl.factorized_joint_logprob({x_rv: x_vv, y_rv: y_vv})
logprob_dict
{x: x_logprob}
```

Reply:
> The bigger problem is that nothing happens if you ask for a graph that we don't know how to handle. That is not specific to censoredRVs, we just haven't tested it. For instance, this snippet does not complain at all:

If our rewrites don't know how to handle something, that's not necessarily a problem.  As a matter of fact, we _expect_ that they won't know how to handle more things than they do.

The assumption underlying your statements and example seems to be that you know what should be done **relative to specific rewrites**, and this is what makes it reasonable to handle rewrite-relevant errors in the rewrite logic.  

In other words, if you know an error/warning should be raised because a value variable specification is redundant, you only really know that because you also know that there's a specific rewrite that determines which value variables are and aren't relevant.

Otherwise, a generic warning for "unused" variable/value mappings is simply an interface choice that might help inform people of issues elsewhere and/or bad assumptions (e.g. that the resulting graph will depend on certain terms), but that's all.
 
All this relates directly to #85.